File: using python/1095. Find in Mountain Array.py
class MountainArray(object):
   def get(self, index):
       """
       :type index: int
       :rtype int
       """

   def length(self):
       """
       :rtype int
       """
class Solution(object):
    def findInMountainArray(self,target, mountain_arr):
        len_arr = mountain_arr.length()
        peak_index = self.find_peak_index(mountain_arr, len_arr)

        # search in the ascending half of the array
        left = 0
        right = peak_index
        while left <= right:
            mid = (left + right) // 2
            if mountain_arr.get(mid) == target:
                return mid
            elif mountain_arr.get(mid) < target:
                left = mid + 1
            else:
                right = mid - 1

        # search in the descending half of the array
        left = peak_index
        right = len_arr - 1
        while left <= right:
            mid = (left + right) // 2
            if mountain_arr.get(mid) == target:
                return mid
            elif mountain_arr.get(mid) < target:
                right = mid - 1
            else:
                left = mid + 1

        return -1  # target not found


    def find_peak_index(self,arr, l):
        left = 0
        right = l - 1
        while left < right:
            mid = (left + right) // 2
            if arr.get(mid) < arr.get(mid + 1):
                left = mid + 1
            else:
                right = mid
        return left

File: Using Java/1095. Find in Mountain Array.java
// Java implementation of the Find in Mountain Array problem
// assuming MountainArray class has get(int index) and length() methods

class Solution {
    public int findInMountainArray(int target, MountainArray mountainArr) {
        int lenArr = mountainArr.length();
        int peakIndex = findPeakIndex(mountainArr, lenArr);

        // search in the ascending half of the array
        int left = 0;
        int right = peakIndex;
        while (left <= right) {
            int mid = (left + right) / 2;
            if (mountainArr.get(mid) == target) {
                return mid;
            } else if (mountainArr.get(mid) < target) {
                left = mid + 1;
            } else {
                right = mid - 1;
            }
        }

        // search in the descending half of the array
        left = peakIndex;
        right = lenArr - 1;
        while (left <= right) {
            int mid = (left + right) / 2;
            if (mountainArr.get(mid) == target) {
                return mid;
            } else if (mountainArr.get(mid) < target) {
                right = mid - 1;
            } else {
                left = mid + 1;
            }
        }

        return -1;  // target not found
    }

    private int findPeakIndex(MountainArray arr, int l) {
        int left = 0;
        int right = l - 1;
        while (left < right) {
            int mid = (left + right) / 2;
            if (arr.get(mid) < arr.get(mid + 1)) {
                left = mid + 1;
            } else {
                right = mid;
            }
        }
        return left;
    }
}
